flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -99,87 +75,360 @@ class DoConcurrentConversion : public mlir::OpConversionPattern<fir::DoLoopOp> {
+                        mlir::Value liveIn) const {
+    hlfir::DeclareOp declareOp =
+        mlir::dyn_cast_if_present<hlfir::DeclareOp>(liveIn.getDefiningOp());
+
+    if (declareOp == nullptr)
+      TODO(liveIn.getLoc(),
+           "Values not defined by declare op's are not supported yet.");
+
+    mlir::Type liveInType = liveIn.getType();
+    mlir::Type eleType = liveInType;
+    if (auto refType = liveInType.dyn_cast<fir::ReferenceType>())
+      eleType = refType.getElementType();
+
+    llvm::omp::OpenMPOffloadMappingFlags mapFlag =
+        llvm::omp::OpenMPOffloadMappingFlags::OMP_MAP_IMPLICIT;
+    mlir::omp::VariableCaptureKind captureKind =
+        mlir::omp::VariableCaptureKind::ByRef;
+
+    if (fir::isa_trivial(eleType) || fir::isa_char(eleType)) {
+      captureKind = mlir::omp::VariableCaptureKind::ByCopy;
+    } else if (!fir::isa_builtin_cptr_type(eleType)) {
+      mapFlag |= llvm::omp::OpenMPOffloadMappingFlags::OMP_MAP_TO;
+      mapFlag |= llvm::omp::OpenMPOffloadMappingFlags::OMP_MAP_FROM;
     }
 
-    // For each collected `fir.store`, find the target memref's alloca's and
-    // declare ops.
-    llvm::SmallSetVector<mlir::Operation *, 4> declareAndAllocasToClone;
-    for (auto storeOp : inductionVarTargetStores) {
-      mlir::Operation *storeTarget = storeOp.getMemref().getDefiningOp();
+    llvm::SmallVector<mlir::Value> boundsOps = [&]() {
+      llvm::SmallVector<mlir::Value> result;
 
-      for (auto operand : storeTarget->getOperands()) {
-        declareAndAllocasToClone.insert(operand.getDefiningOp());
+      if (declareOp.getShape() == nullptr) {
+        return result;
       }
-      declareAndAllocasToClone.insert(storeTarget);
+
+      fir::ShapeOp shapeOp = mlir::dyn_cast_if_present<fir::ShapeOp>(
+          declareOp.getShape().getDefiningOp());
+
+      if (shapeOp == nullptr)
+        TODO(liveIn.getLoc(),
+             "Shapes not defined by shape op's are not supported yet.");
+
+      auto extents = shapeOp.getExtents();
+
+      auto genBoundsOp = [&](mlir::Value extent) {
+        mlir::Type extentType = extent.getType();
+        mlir::arith::ConstantOp lb = rewriter.create<mlir::arith::ConstantOp>(
+            liveIn.getLoc(), extentType,
+            rewriter.getIntegerAttr(extentType, 0));
+        // TODO I think this caluclation might not be correct. But this is how
+        // it is done in PFT->OpenMP lowering. So keeping it like this until we
+        // double check.
+        mlir::Value ub =
+            rewriter.create<mlir::arith::SubIOp>(liveIn.getLoc(), extent, lb);
+
+        return rewriter.create<mlir::omp::MapBoundsOp>(
+            liveIn.getLoc(), rewriter.getType<mlir::omp::MapBoundsType>(), lb,
+            ub, extent, mlir::Value{}, false, mlir::Value{});
+      };
+
+      for (auto extent : extents)
+        result.push_back(genBoundsOp(extent));
+
+      return result;
+    }();
+
+    return Fortran::lower::omp::createMapInfoOp(
```

Comment:
The map code looks reasonable for the moment! I would just advise that you make sure this pass occurs prior to the OpenMP descriptor pass otherwise the descriptor types (allocatables et al) that are captured won't be expanded appropriately which will result in them not being mapped correctly. I think that should be enough to have these types mapped implicitly.

You probably already have the passes pipelined like this but just wanted to make sure!

Reply:
This little segment of code will also need to maintain lockstep with any changes to the implicit mapping in the OpenMP lowering as well unfortunately. There is nothing new yet, but there likely will be in the future but I can perhaps update this or keep you informed of any changes I make (or I see being made) :-)